### PR TITLE
Make --symlink idempotent

### PR DESCRIPTION
**Pull request by jkaivo**
When creating symlinks specified with --symlink, check errno if creation fails. If the reason was EEXIST, check the path of the existing symlink. If it matches what was desired, continue as normal.

### CODE REVIEW

**Comment by rusty-snake** [comment]
Fixes #549

**Review by smcv — CHANGES_REQUESTED** [review]
Thanks for contributing this! I don't know whether it's OK for this to be unconditional, or whether that would be a breaking change; that can be a separate thread. For now I'm reviewing the implementation as if it's OK to do this simple version.

**Comment by smcv** [comment]
@alexlarsson, @cgwalters: Do you think it's OK to make `--symlink` idempotent unconditionally, or do you think there's a risk of breaking existing `bwrap` callers that are expecting the usual atomic "if exists then error else create" semantics?

If you think it needs to be explicit, there are a couple of possible spellings.

The simplest but least general would be `--symlink-idempotent desired-target /dest`, like the relationship between `--ro-bind` and `--bind`.

But I think it might be better to do something more like the recently-added `--perms`, like maybe these (which would not all have to be implemented immediately, but it could be useful to have syntax available for them):

* `--conflict exclusive --symlink t /dest`: if `/dest` already exists then error, even if it's a symlink to `t`
* `--conflict idempotent --symlink t /dest`: if `/dest` already exists with the desired target then ignore, else error
* `--conflict rename --symlink t /dest`: if `/dest` already exists and doesn't have the desired target, then try to rename it to `/dest.old.{unique string}`, and if successful, leave it there
* `--conflict rename-or-remove --symlink t /dest`: if `/dest` already exists and doesn't have the desired target, then try to rename it to `/dest.old.{unique string}`; if successful and empty, remove the old name; if successful and non-empty, keep it

**Comment by smcv** [comment]
If we had a `--conflict` option acting as a modifier for the next action, then we'd be able to expand that to other file types over time (for example `--dir` currently defaults to the equivalent of `--conflict idempotent --dir`). Obviously implementing that wouldn't be a blocker for this feature: it would be fine for an initial implementation of `--conflict` to only work on `--symlink`.

File: bubblewrap.c
Comment (jkaivo) [reply bubblewrap.c:1478]: Fair enough. I didn't see a coding style explicitly called out anywhere, but I'll run it through indent with GNU options.

File: bubblewrap.c
Comment (jkaivo) [reply bubblewrap.c:1481]: Sounds reasonable.

File: bubblewrap.c
Comment (jkaivo) [reply bubblewrap.c:1487]: Simple enough.

File: bubblewrap.c
Comment (jkaivo) [reply bubblewrap.c:1482]: Same as above.

File: bubblewrap.c
Comment (cgzones) [inline bubblewrap.c:1481]: `readlink`(2) does not guarantee to NUL terminate on success

Comment (jkaivo) [reply bubblewrap.c]: Simple enough to account for.

**Review by AndersBlomdell — APPROVED** [review]
A very nice enhancement to flatpak, please merge.

Maybe after fixing the whitespace problem on line 1493 (two extra blanks before break)

**Comment by smcv** [comment]
I was really hoping for a review of this from @cgwalters or @alexlarsson, who have a better overview than I do of all the places where bubblewrap is used and the ways in which it needs to stay backwards-compatible.

> @jchernand0 approved these changes

> @AndersBlomdell approved these changes

Did I miss one of the maintainers of bubblewrap adding you to the project?

Or if you are not maintainers, does this mean we can count on receiving contributions from you if this change is merged but turns out to cause regressions?

File: bubblewrap.c
Comment (smcv) [inline bubblewrap.c:1490]: Please use `readlink_malloc()` (into a new `cleanup_free char *existing = NULL`) in preference to a statically-sized buffer. This returns a non-null string on success, or `NULL` with `errno` set on failure.

**Review by smcv — CHANGES_REQUESTED** [review]

**Comment by AndersBlomdell** [comment]
> I was really hoping for a review of this from @cgwalters or @alexlarsson, who have a better overview than I do of all the places where bubblewrap is used and the ways in which it needs to stay backwards-compatible.
> 
> > @jchernand0 approved these changes
> 
> > @AndersBlomdell approved these changes
> 
> Did I miss one of the maintainers of bubblewrap adding you to the project?
> 
> Or if you are not maintainers, does this mean we can count on receiving contributions from you if this change is merged but turns out to cause regressions?

Not a maintainer, but since this patch actually made flatpak useful for me (I add symlinks below /run/user), I will of course help to the best of my abilities.

**Comment by cwajh** [comment]
@smcv If you need an LGTM from @alexlarsson or @cgwalters before you'll feel comfortable accepting this pull request, what do we have to do to get their attention? The pull request has been waiting for them since April and I'm sure @jkaivo would be more motivated to implement your comment if we had some indication that the patch might actually go in once the fix is made.

If the concern is backwards-incompatibility with current users... I for one can testify that certain users of bwrap are already _assuming_ the functionality is idempotent, and choking when it isn't. That's the reason I found this thread in the first place: a flatpak app I use fails to start if one of the many hidden directories in my homedir has a symlink to somewhere. I'm working around it with `mount --bind` but it's fairly annoying to need special root jiggery-pokery to do something that should be a simple `ln`.

Thanks for your help.

File: bubblewrap.c
Comment (AleVul) [reply bubblewrap.c:1490]: hey @jkaivo will you be able to do this chages?

Comment (jkaivo) [reply bubblewrap.c:1490]: I'm just seeing this. I don't see what actual advantage `readlink_malloc()` has over POSIX `readlink()` with an appropriately sized buffer. Since submitting this patch, I have solved my personal use-case issues by only using Steam on my Steam Deck, so I have very little motivation to make additional changes.

**Comment by alexlarsson** [comment]
Personally I think it is pretty safe to just add this as-is. I'm not aware of anyone using bwrap --symlink and then specifically looking for some kind of EEXIST error and handling that in a custom way. It would be really hard to extract that particular error anyway as other things could error out.

So, LGTM, but I would prefer if all the changes were squashed into one commit.

File: bubblewrap.c
Comment (smcv) [reply bubblewrap.c:1490]: > I don't see what actual advantage readlink_malloc() has over POSIX readlink() with an appropriately sized buffer

If it was possible to choose an appropriately sized buffer, then there would be no advantage. The problem with POSIX readlink() is that there is no correct size for the buffer, because Linux `PATH_MAX` is a lie (it is entirely valid to have paths longer than `PATH_MAX`).

**Comment by smcv** [comment]
It sounds as though the submitter of this PR is no longer interested in contributing it, so I'll try to find time among my other responsibilities to rebase this and fix my review comment myself.

**Comment by AndersBlomdell** [comment]
> It sounds as though the submitter of this PR is no longer interested in contributing it, so I'll try to find time among my other responsibilities to rebase this and fix my review comment myself.

Something along these lines?:

```
diff --git a/bubblewrap.c b/bubblewrap.c
index e894af7..65e5563 100644
--- a/bubblewrap.c
+++ b/bubblewrap.c
@@ -1468,7 +1468,24 @@ setup_newroot (bool unshare_pid,
         case SETUP_MAKE_SYMLINK:
           assert (op->source != NULL);  /* guaranteed by the constructor */
           if (symlink (op->source, dest) != 0)
-            die_with_error ("Can't make symlink at %s", op->dest);
+            {
+              if (errno == EEXIST)
+                {
+                  cleanup_free char *existing = readlink_malloc(dest);
+                  if (existing == NULL) {
+                    if (errno == EINVAL)
+                      die ("Can't make symlink at %s: destination exists and is not a symlink", op->dest);
+                    else
+                      die_with_error ("Can't make symlink at %s: destination exists, and cannot read symlink target", op->dest);
+                  }
+
+                  if (strcmp (existing, op->source) == 0)
+                    break;
+
+                  die ("Can't make symlink at %s: existing destination is %s)", op->dest, existing);
+                }
+              die_with_error ("Can't make symlink at %s", op->dest);
+            }
           break;
 
         case SETUP_SET_HOSTNAME:
```

**Comment by smcv** [comment]
> Something along these lines?:

Yes, a lot like that (but with some coding style fixes). Thanks!

Please send pull requests for any future changes that you want to propose: the limiting factor on this project is the  maintainers' limited time, and that's more efficiently used if we aren't passing diff output around like it's the 1990s.

> `+                  cleanup_free char *existing = readlink_malloc(dest);`

coding style: space before `(`

> `+                  if (existing == NULL) {`

coding style: this project uses GNU-style indentation

> `+                  die ("Can't make symlink at %s: existing destination is %s)", op->dest, existing);`

The trailing `)` looks wrong here

**Review by smcv — APPROVED** [review]

**Comment by smcv** [comment]
I'll merge this after CI succeeds, unless another maintainer objects.

**Pull request merged**
